src/main/common/pwl.c

```diff
@@ -0,0 +1,63 @@
+/*
+ * This file is part of Betaflight.
+ *
+ * Betaflight is free software. You can redistribute this software
+ * and/or modify this software under the terms of the GNU General
+ * Public License as published by the Free Software Foundation,
+ * either version 3 of the License, or (at your option) any later
+ * version.
+ *
+ * Betaflight is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ *
+ * See the GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this software.
+ *
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "platform.h"
+
+#include "pwl.h"
+
+
+void pwlInitialize(pwl_t *pwl, float *yValues, int numPoints, float xMin, float xMax) {
+    pwl->numPoints = numPoints;
+    pwl->xMin = xMin;
+    pwl->xMax = xMax;
+    pwl->dx = (xMax - xMin) / (numPoints - 1);
+    pwl->yValues = yValues;
+}
+
+void pwlFill(pwl_t *pwl, float (*function)(float))
+{
+    for (int i = 0; i < pwl->numPoints; ++i) {
+        const float x = pwl->xMin + i * pwl->dx;
+        pwl->yValues[i] = function(x);
+    }
+}
+
+float pwlInterpolate(const pwl_t *pwl, float x)
+{
+    if (x <= pwl->xMin) {
```

Comment:
(another possibility, probably over-optimized. Main advantage is that dx is not necessary)

```
{
    float t = (x -  pwl->xMin) / (pwl->xMax - pwl->xMin);  // map to xMin - xMax interval
    float iptrf;
    float frac = modff(t * (pwl->numPoints - 1), &iptrf); // split to integer and fractional part
    int index =  iptrf;
    if (index < 0) {
       return return pwl->yValues[0];
    }
    if (index >= pwl->numPoints - 1) {
        return pwl->yValues[pwl->numPoints - 1];
    }
    const float y0 = pwl->yValues[index];
    const float y1 = pwl->yValues[index + 1];

    return y0 + frac * (y1 - y0);
}

Reply:
hmmm. based on some performance tests i just did, the frac-modff version is consistently 10-25 % slower. Here is the performance test code, pretty basic:
[ConsoleApplication1.c.txt](https://github.com/user-attachments/files/16515072/ConsoleApplication1.c.txt)
Or here in the online compiler:
https://www.programiz.com/online-compiler/4wQBwoPChTWbI

But i also checked locally - same thing.

Reply:
It can be optimized a bit:
```
    float t = (x -  pwl->xMin) / pwl->dx;  // map to xMin .. xMax to 0 ..  (pwl->numPoints - 1) interval
    float iptrf;
    float frac = modff(t, &iptrf); // split to integer and fractional part
```

But don't worry about it, you version is fine. I only have small problem with numPoints / xMax / dx redundancy, only two of them are necessary (and rounding or other small errors may theoretically lead to problems)